APIv3 attachments

NobodysNightmare opened the pull request:
## OpenProject work packages
- EPIC: https://community.openproject.org/work_packages/19339
- Story: https://community.openproject.org/work_packages/19988
## Description

This PR adds and alters endpoints around attachments. The following things should become possible with this PR:
- [x] obtaining metadata about an attachment
- [x] download an attachment
- [x] list attachments of a work package
- [x] add an attachment to a work package
- [x] delete an attachment

Most parts of the specification are already done and feedback is welcome.
## Further TODOs
- [x] integrative specs for attachment listing
- [x] validate inputs on attachment upload
- [x] allow empty description
- [x] add specs for attachment upload
- [x] check how/when file size constraints are applied (and how to catch them)
- [x] also link attachments in WP
- [x] update docs with 400 responses of attachment upload
- [x] DRY up attachment upload code ~~(use anything built-in for JSON-parsing + error handling)~~

NobodysNightmare committed:
specify the attachments resource

NobodysNightmare committed:
specify attachment listing per work package

NobodysNightmare committed:
fix spelling

NobodysNightmare committed:
rework attachments endpoint according to spec

- drop useless properties
- add essential properties
- for now limit to work packages only

NobodysNightmare committed:
style docs

- unify headings
- some wording
- add missing heading

NobodysNightmare committed:
add flavour text

NobodysNightmare committed:
fix documentation errors

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-apiv3-attachments.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-apiv3-attachments.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
specify file upload

NobodysNightmare committed:
delete attachments via APIv3

NobodysNightmare committed:
list attachments of a WP via APIv3

NobodysNightmare committed:
style

NobodysNightmare committed:
specs for attachment listing

NobodysNightmare committed:
expect files to be uploaded as multipart/form-data

instead of using the (as far as I could find out) more apropriate multipart/related.

It looks like grape does not support multipart/related apropriately

NobodysNightmare committed:
make attachment upload WORK

lets see if we can make it GOOD, I think it already is FAST ^^

NobodysNightmare committed:
correct (but ugly) upload error handling

NobodysNightmare committed:
ensure correct file size errors are displayed

NobodysNightmare committed:
happy path spec for attachment upload

NobodysNightmare committed:
replace create_uploaded_file in specs

mock_uploaded_file creates a Rack::Test::UploadedFile, which should behave equally to a real uploaded file,
but is also treated correctly by Rack::Test

NobodysNightmare committed:
add specs for the unhappy pathes

those are quite a lot of specs for an integrative test
however at the moment there is (too) much logic in the endpoint itself, so this seems to be neccessary

NobodysNightmare committed:
some cleanup

I just extracted some methods... the basic mess is the same, but hidden behind some methods...

think of a pig with make up, it looks good, but it is still a pig.

NobodysNightmare committed:
link attachments in WP

for now only a link without embedding

NobodysNightmare committed:
unify link specs for WP representer

- use shared examples when apropriate
- move shared examples from 'links.rb' to 'link_examples.rb'

NobodysNightmare committed:
update specs to include error responses

NobodysNightmare commented on lib/api/v3/attachments/attachments_by_work_package_api.rb:
```diff
@@ -0,0 +1,92 @@
+
+            unless metadata && file
+              raise ::API::Errors::InvalidRequestBody.new(
+                      I18n.t('api_v3.errors.multipart_body_error'))
```
rubocop is not offended, so you are either talking outdated little hound or you are wrong...

NobodysNightmare commented on lib/api/v3/attachments/attachments_by_work_package_api.rb:
```diff
@@ -0,0 +1,106 @@
+
+              unless file_name
+                raise ::API::Errors::Validation.new(
+                        "fileName #{I18n.t('activerecord.errors.messages.blank')}.")
```
nope

> 1 file inspected, no offenses detected

You are wrong, Sir!

NobodysNightmare committed:
fix style issues

NobodysNightmare commented on lib/api/v3/utilities/path_helper.rb:
```diff
@@ -46,6 +46,14 @@ def self.attachment(id)
 
+          def self.attachment_download(id)
+            "#{root_path}attachments/#{id}"
+          end
```
RFC: I could also exclude this link form the APIv3 path helper and use the helpers of rails instead.

At this point I am unsure how to balance rails-independence vs. code repetition...

ulferts replied:
Why not use the Rails path inside this method here so that you avoid the repetition but have a consistent path registry for the whole of the API.

NobodysNightmare replied:
:+1: 

Best of both worlds, I guess...

ulferts commented on config/locales/de.yml:
```diff
@@ -1695,6 +1697,8 @@ de:
       invalid_relation: "Die Beziehung ist ungültig."
       invalid_resource: "Für die Eigenschaft '%{property}' wird ein Link wie '%{expected}' erwartet. Gesandt wurde '%{actual}'."
       missing_content_type: "nicht angegeben"
+      invalid_json: "The request could not be parsed as JSON."
```
This is not at it's proper place. I18n should be in alphabetical order.

ulferts commented on lib/open_project/files.rb:
```diff
@@ -56,5 +56,14 @@ def create_uploaded_file(name: 'test.txt',
+
+      tmp = create_temp_file name: name, content: content, binary: binary
+      Rack::Test::UploadedFile.new tmp.path, content_type, binary
+    end
```
I think there is a better place for code you require only in your tests than in the production code.

NobodysNightmare replied:
Tell this the guy writing `create_uploaded_file`...

I did not verify that the only productive occurence is valid after all ^^

But you are right, existing bad code is no excuse for producing new one ^^

ulferts commented on lib/api/v3/attachments/attachments_by_work_package_api.rb:
```diff
@@ -0,0 +1,106 @@
+            end
+
+            ::API::V3::Attachments::AttachmentRepresenter.new(attachment)
+          end
```
This class is really cluttered by all the hacks you have too pull of in order to make it work. Any chance we can put those hacks into a separate class so that the controller only has to deal with the request/response flow again?

NobodysNightmare replied:
@ulferts I tried to refactor the class a little more.

I think it should now have returned to readability again, tell me what you think.

ulferts commented:
This of course breaks the front end but @NobodysNightmare already told you that, right @0xF013?

ulferts commented:
Functionality seems to work fine. The code isn't beautiful but I don't have easy recommendations to fix this. So apart from my remarks and once the front end is working again, this could be good to merge.

NobodysNightmare committed:
ask the rails url helper for the attachment DL URL

NobodysNightmare committed:
sorting some localization properties

NobodysNightmare committed:
move mock method out of production code

NobodysNightmare commented:
@0xF013 Don't worry, I will handle the frontend, can't be that hard, right? ^^

edit: It still sucks to break stuff and have no tests failing...

0xF013 commented:
@NobodysNightmare yeah, just use some jquery plugins

NobodysNightmare committed:
parse attachment metadata using a representer

- adding general error handling for parse errors
- changing the default error message for parse errors, which was very form specific before

NobodysNightmare committed:
move uploaded file creation out of api endpoint

NobodysNightmare committed:
extract method

NobodysNightmare committed:
please the hound

NobodysNightmare committed:
always embed the attachments

ulferts commented:
@NobodysNightmare ping, specs are failing.

ulferts commented on lib/api/v3/attachments/attachment_metadata_representer.rb:
```diff
@@ -0,0 +1,47 @@
+      end
+    end
+  end
+end
```
Sorry to pester you because I really appreciate you moving the stuff from the controller to the representer but this one here lacks specs.

NobodysNightmare replied:
Pestering is a good thing... Apparently there was a bug I introduced with the representer, but the specs found it ^^

It was not possible to set a description.

ulferts commented:
@NobodysNightmare thanks for the cleanup work. And kudos for getting the front end working again :bow: Now that you can do stuff like this, you can probably support @0xF013. There are some tasks that are needed urgently.

I think this is good to merge once the tests are fixed.

NobodysNightmare commented:
> Now that you can do stuff like this, you can probably support @0xF013

Actually I hoped to make some glorious mistakes so that nobody would ever let me touch the frontend again, but that plan probably failed ^^

NobodysNightmare committed:
fix all the legacy specs

NobodysNightmare committed:
fix karma tests

NobodysNightmare committed:
let karma test stubs look more like the real thing

NobodysNightmare committed:
add specs for metadata representer

NobodysNightmare commented on frontend/tests/unit/tests/work_packages/controllers/work-package-details-controller-test.js:
```diff
@@ -82,7 +87,7 @@ describe('WorkPackageDetailsController', function() {
         },
         links: {
-          self: "it's a me, it's... you know...",
+          self: { href: "it's a me, it's... you know..." },
```
I think you are drunk

NobodysNightmare commented:
@ulferts Any reason for the work in progress label?

I was actually done and the only thing missing 20 hours ago were probably travis tests. I restarted two flickering jobs just now.
